libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/WebSocketEnabledHttpAdapter.cs

```diff
@@ -0,0 +1,179 @@
+        /// <param name="credentialProvider">Optional credential provider.</param>
+        /// <param name="channelProvider">Optional channel provider.</param>
+        /// <param name="loggerFactory">Optional logger factory.</param>
+        public WebSocketEnabledHttpAdapter(IConfiguration configuration, ICredentialProvider credentialProvider = null, IChannelProvider channelProvider = null, ILoggerFactory loggerFactory = null)
+            : this(configuration, null, credentialProvider ?? new ConfigurationCredentialProvider(configuration), channelProvider ?? new ConfigurationChannelProvider(configuration), loggerFactory)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WebSocketEnabledHttpAdapter"/> class.
+        /// An adapter meant for subclassing and injecting custom implementation of BotFrameworkHttpAdapter for handling Http requests.
+        /// Throws <see cref="ArgumentNullException"/> if configure is null.
+        /// </summary>
+        /// <param name="configuration">The configuration for the adapter to use.</param>
+        /// <param name="botFrameworkHttpAdapter">Optional http adapter to use for non-streaming extensions requests.</param>
+        /// <param name="credentialProvider">Optional credential provider.</param>
+        /// <param name="channelProvider">Optional channel provider.</param>
+        /// <param name="loggerFactory">Optional logger factory.</param>
+        protected WebSocketEnabledHttpAdapter(IConfiguration configuration, IBotFrameworkHttpAdapter botFrameworkHttpAdapter = null, ICredentialProvider credentialProvider = null, IChannelProvider channelProvider = null, ILoggerFactory loggerFactory = null)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            var openIdEndpoint = configuration.GetSection(AuthenticationConstants.BotOpenIdMetadataKey)?.Value;
+
+            if (!string.IsNullOrEmpty(openIdEndpoint))
+            {
+                // Indicate which Cloud we are using, for example, Public or Sovereign.
+                ChannelValidation.OpenIdMetadataUrl = openIdEndpoint;
+                GovernmentChannelValidation.OpenIdMetadataUrl = openIdEndpoint;
+            }
+
+            credentialProvider = credentialProvider ?? new ConfigurationCredentialProvider(configuration);
+            channelProvider = channelProvider ?? new ConfigurationChannelProvider(configuration);
+
+            _botFrameworkHttpAdapter = _botFrameworkHttpAdapter ?? new BotFrameworkHttpAdapter(credentialProvider, channelProvider, loggerFactory?.CreateLogger<BotFrameworkHttpAdapter>());
+            _webSocketConnector = new WebSocketConnector(credentialProvider, channelProvider, loggerFactory?.CreateLogger<WebSocketConnector>());
+
+            _ensureMiddlewareSet = new Lazy<bool>(() =>
+            {
```

Comment:
I don't unerstand this code or the need for it. Can we remove it? If this is due to a bug somewhere else, let's fix that bug.

Reply:
This is a workaround until there's a fix for https://github.com/microsoft/botbuilder-dotnet/issues/2138